reflect/src/main/java/dagger/reflect/ReflectiveMembersInjector.java

```diff
@@ -87,43 +87,43 @@
       }
 
       if (!fieldProviders.isEmpty() || !methodProviders.isEmpty()) {
-        // [@Inject] Fields and methods in superclasses are injected before those in subclasses.
-        // we are walking up (getSuperclass), but spec says @Inject should be walking down the hierarchy
-        hierarchyBindings.addFirst(new ClassBindings(fieldProviders, methodProviders));
+        // Per JSR 330, fields and methods in superclasses are injected before those in subclasses.
+        // We are traversing upward in the class hierarchy so each injector is prepended to the
+        // collection to ensure regular iteration will honor this contract.
+        classInjectors.addFirst(new ClassInjector<>(fieldProviders, methodProviders));
       }
 
       target = target.getSuperclass();
     }
 
-    return new ReflectiveMembersInjector<>(hierarchyBindings);
+    return new ReflectiveMembersInjector<>(classInjectors);
   }
 
-  private final Iterable<ClassBindings> classBindings;
+  private final Iterable<ClassInjector<T>> classInjectors;
 
-  private ReflectiveMembersInjector(Iterable<ClassBindings> classBindings) {
-    this.classBindings = classBindings;
+  private ReflectiveMembersInjector(Iterable<ClassInjector<T>> classInjectors) {
+    this.classInjectors = classInjectors;
   }
 
   @Override public void injectMembers(T instance) {
-    for (ClassBindings classBinding : classBindings) {
-      classBinding.injectMembers(instance);
+    for (ClassInjector<T> classInjector : classInjectors) {
+      classInjector.injectMembers(instance);
     }
   }
 
-  private static final class ClassBindings {
+  private static final class ClassInjector<T> implements MembersInjector<T> {
     final Map<Field, Provider<?>> fieldProviders;
     final Map<Method, Provider<?>[]> methodProviders;
 
-    ClassBindings(
+    ClassInjector(
         Map<Field, Provider<?>> fieldProviders,
         Map<Method, Provider<?>[]> methodProviders) {
       this.fieldProviders = fieldProviders;
       this.methodProviders = methodProviders;
     }
 
-    void injectMembers(Object instance) {
-      // [@Inject] Constructors are injected first, followed by fields, and then methods.
-      // Note: the constructor injection is in dagger.reflect.Binding.UnlinkedJustInTime.dependencies
+    @Override public void injectMembers(T instance) {
+      // Per JSR 330, fields are injected before methods.
```

Comment:
Where did you see this? `Inject` docs say 
> Ordering of injection among fields and among methods in the same class is not specified.

Reply:
https://docs.oracle.com/javaee/6/api/javax/inject/Inject.html

> among fields and among methods in the same class

this means that the `fieldProviders` and `methodProviders` iteration order can be unspecified, which it is, because `getDeclaredFields` and `getDeclaredMethods` doesn't guarantee any ordering.

this comment refers to the previous sentence in Inject javadoc:
> Constructors are injected first, followed by fields, and then methods. Fields and methods in superclasses are injected before those in subclasses.

and it specifies that `fieldProviders` must be iterated before `methodProviders`